### ISSUE TITLE
`__iter__` on subclass iterates over plain `Tensor`s instead of instances of the subclass

### ISSUE THREAD

**Issue by kosiokarchev** [pytorch/pytorch#54457]
As far as I understand the latest subclassing semantics, all operations on subclasses of `Tensor` are supposed to return instances of the subclass. However, it seems like `__iter__` returns an iterator over plain `Tensor` instances instead of the subclass. A simple example would be

```
import torch

class SubTensor(torch.Tensor):
    pass

t = SubTensor(torch.empty(1, 5))

print(type(t[0]))           # <class '__main__.SubTensor'> as expected
print(type(next(iter(t))))  # <class 'torch.Tensor'>
```

If this is the intended behaviour, what would be the way to override it beside manually setting the `__class__` on each element of the iterator?

cc @hameerabbasi @rgommers

**Comment by ezyang** [comment]
I agree with your reasoning, this is a bug. Instead of setting `__class__`, I think you can also use `torch.Tensor._make_subclass`.

**Comment by hameerabbasi** [comment]
> I agree with your reasoning, this is a bug. Instead of setting `__class__`, I think you can also use `torch.Tensor._make_subclass`.

I think `as_subclass` is the better choice.

**Comment by hameerabbasi** [comment]
Okay, this is a weird one, and can be solved by _removing_ `__torch_function__` dispatch from `Tensor.__iter__`.

### What happens exactly
`Tensor.__iter__` is called with `self` being an instance of `SubTensor`. it goes through the dispatch properly, and comes back into `__iter__` a second time as designed with dispatch turned off, returning an iterator of a `tuple`. This iterator isn't properly detected as containing `Tensor`s, and therefore just passes through "unconverted", therefore returning the original `Tensor`s.

### Possible Fix 1 (Easy)
Remove `__torch_function__` from `__iter__`, it will work through `__torch_function__` on `unbind`. We just declare it as "one of those things" where using `unbind` will do the right thing automagically.

### Possible Fix 2 (Hard)
Declare a custom iterator class/generator that is a catch-all for iterators, converting each element.

**Comment by ezyang** [comment]
Easy fix sounds easy, and we already have "those things"

**Comment by hameerabbasi** [comment]
I'm looking into this in #54702.

**Comment by hameerabbasi** [comment]
This was fixed in #54702.
